### PR TITLE
Fix #15166: Only use custom stem directions when stem is unbeamed

### LINKED ISSUE #15166: [MU4 Issue] Broken stems in drum score
**Describe the bug**
In a drum score, trying to write a) bass drum, pedal hi-hat or high floor tom and b) other parts of the drum set in one voice while otherwise using MuseScore defaults (stem directions downward for bass drum, pedal hi-hat note, high floor tom) results in broken stems. This happens both when entering notes in a newly created drum score in MuseScore 4 and when opening a drum score created in MuseScore 3 in MuseScore 4.

By default, MuseScore's drum input places bass drum, pedal hi-hat and high floor tom in the second voice but in drum notation it is also common to place all parts of the drum set in one voice.

**To Reproduce**
1. Create a drum score in MuseScore 4.
2. Write consecutive beamed notes (16ths, quavers, …) of bass drum and snare drum hits with both placed in the same voice.

or
1. Open the attached score in MuseScore 4: [CultOfPersonality-IntroFill.zip](https://github.com/musescore/MuseScore/files/10228439/CultOfPersonality-IntroFill.zip)
2. See the broken stems in bars 1, 3 and 5.

**Expected behavior**
No broken stems

**Screenshots**
MuseScore 3: MuseScore3
MuseScore 4: MuseScore4

**Platform information**
 - OS: macOS 11.7.2, final release of MuseScore 4

### PR DESCRIPTION
**Pull request by asattely**
Resolves: https://github.com/musescore/MuseScore/issues/15166
A better fix for this one (than PR https://github.com/musescore/MuseScore/pull/15205, which treats it as an import issue) is to just ignore explicit stem direction if the chord is beamed. This will allow stem directions to flip when the beam is manually adjusted, as well as weird drum situations where stem directions are hardcoded even when beamed in a certain direction.

(partially) Resolves: https://github.com/musescore/MuseScore/issues/15298
There is an unrelated issue where converting a chord from a normal chord to a tuplet resets its stem direction to `AUTO`, which is not solved by this PR. However, when other chords are added to the tuplet in a percussion staff, the added chords have explicit stem direction so the beam will use that. Basically, when determining the direction of the beam, it looks through all the chords to see if there are any explicit stem directions, and if so, uses that.

Chords not maintaining their explicit stem direction when tupletified is fixed by https://github.com/musescore/MuseScore/pull/15394.

Resolves: https://github.com/musescore/MuseScore/issues/15401
Determining stem direction for notes within beams was thrown off when the beam both started and ended with a rest. Now the direction calculation for the notes within a beam match the calculation for the beam itself--the first and last chords of a beam are determined by ignoring the rests and using the first and last actual chords of a beam

### CODE REVIEW

**Review by oktophonie — APPROVED** [review]

**Comment by asattely** [comment]
> Does this relate to https://musescore.org/en/node/264750 ?
It does indeed. As long as the chords have explicit stem direction, the beam will abide by that.

**Review by oktophonie — APPROVED** [review]

**Comment by Eism** [comment]
@asattely 
Please fix utests

**Comment by asattely** [comment]
Rebased, utests good!

File: src/engraving/libmscore/beam.cpp [hunk 1]
```diff
@@ -318,11 +318,22 @@ void Beam::layout1()
...
+            // search through the beam for the first chord with explicit stem direction and use that.
+            // if there is no explicit stem direction, default to the direction of the first stem.
+            bool firstUp = false;
+            bool firstChord = true;
+            DirectionV explicitDirection = DirectionV::AUTO;
             for (ChordRest* cr :_elements) {
                 if (cr->isChord()) {
-                    _up = cr->up();
-                    break;
+                    if (toChord(cr)->stemDirection() != DirectionV::AUTO) {
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/beam.cpp:328]: Please, never call the same operations multiple times even if they are relatively fast to perform. Call this stemDirection() once, write its result to a temp variable, and reuse it

**Comment by RomanPudashkin** [comment]
> Could you please add some utests for Chord::computeUp()?

@asattely Any progress with this?

**Comment by RomanPudashkin** [comment]
> I feel that the functionality is better tested via a vtest, since it just loads the file and makes sure the stem directions laid out correctly based on the new computeUp functionality.

Could you please rebase your PR? We will see if the vtests show the difference. If they are all "green" after rebasing, then we should add more tests to cover all possible cases

**Review by oktophonie — APPROVED** [review]

**Review by RomanPudashkin — APPROVED** [review]

**Pull request merged**